Add support for SQL

bohan-zhang opened the pull request:
@hartikainen This is my progress so far. However, I am encountering an issue with policies:
```
Traceback (most recent call last):
  File "/Users/bohan/miniconda2/envs/softlearning/lib/python3.6/site-packages/ray/tune/function_runner.py", line 80, in run
    self._entrypoint(*self._entrypoint_args)
  File "/Users/bohan/Documents/EngSci/4-1/research/softlearning/examples/multi_goal/main.py", line 53, in run_experiment
    plotter=plotter,
  File "/Users/bohan/Documents/EngSci/4-1/research/softlearning/softlearning/algorithms/utils.py", line 36, in get_algorithm_from_variant
    variant, *args, **algorithm_kwargs, **kwargs)
  File "/Users/bohan/Documents/EngSci/4-1/research/softlearning/softlearning/algorithms/utils.py", line 15, in create_SQL_algorithm
    algorithm = SQL(*args, **kwargs)
  File "/Users/bohan/Documents/EngSci/4-1/research/softlearning/softlearning/algorithms/sql.py", line 117, in __init__
    self._create_svgd_update()
  File "/Users/bohan/Documents/EngSci/4-1/research/softlearning/softlearning/algorithms/sql.py", line 219, in _create_svgd_update
    actions = self.policy.actions_for(
AttributeError: 'FeedforwardGaussianPolicy' object has no attribute 'actions_for'
```

I've been trying to use `actions_np`, but it seems to have other issues.

bohan-zhang commented:
@hartikainen @haarnoja I'm still stuck on `_create_svgd_update` in `sql.py`. I tried using `actions` to replace `actions_for`, but if you can help me out on that part, that'll save me so much time!

hartikainen commented:
Hey @bohan-zhang. Thanks a lot for the contribution! I'm gone for couple of days, and will get back on this on Wednesday.

bohan-zhang commented:
@hartikainen enjoy your break and I’ll try to figure this out to some extent!

bohan-zhang commented on softlearning/algorithms/sql.py:
```diff
@@ -216,10 +217,10 @@ def _create_td_update(self):
-            reuse=True)
+        action_tensors = [tf.expand_dims(self.policy.actions(self._observations_ph), -2)
+                          for _ in range(self._kernel_n_particles)]
+        actions = tf.concat(action_tensors, 1)
```
@hartikainen these fixes are very hack-y. However, I haven't come up with a way to fix it.

bohan-zhang replied:
This doesn't seem to work :(

hartikainen replied:
When you say it doesn't work, does it crash or does it not learn?

hartikainen replied:
Seems like this works fine. I pushed this over to your branch.

bohan-zhang commented on softlearning/models/feedforward.py:
```diff
@@ -27,7 +27,7 @@ def feedforward_model(input_shapes,
 
     concatenated = tf.keras.layers.Lambda(
-        lambda x: tf.concat(x, axis=-1)
+        lambda x: tf.concat(x, axis=-2)
```
@hartikainen in some cases, we need to concatenate `[None, 1, 2]` shape with `[None, 8, 2]`, so we need some flexibility to account for `axis` argument.

bohan-zhang commented:
@hartikainen I made more progress but I'm not entirely sure how to replace the `self.policy.get_params_internal()` function. Also made some comments in the commits.

hartikainen commented:
@bohan-zhang I pushed couple of commits, mainly to fix the `_create_svgd_update`. I didn't yet test if the algorithm fully works, but at least I was able to get the multi_goal example running.

I would revert the change in the `feedforward.py`, and possibly change the way actions are fetched in the beginning of `_create_svgd_update` (see my comments inline).

Good job! Let me know if there's anything else you need.

hartikainen commented on softlearning/algorithms/sql.py:
```diff
@@ -236,58 +239,57 @@ def _create_svgd_update(self):
         gradients = tf.gradients(
             updated_actions,
-            self.policy.get_params_internal(),
+            self.policy.trainable_variables,
```
Re your questions about the policy trainable variables. I previously changed our policy (and value function) apis to use `trainable_variables` property instead of `get_params_internal` method. I think this is better in line with the standard way of doing things, at least in keras.

bohan-zhang commented:
#20

hartikainen commented on examples/development/variants.py:
```diff
@@ -90,9 +105,9 @@
-                    MAX_PATH_LENGTH_PER_DOMAIN.get(
-                        domain, DEFAULT_MAX_PATH_LENGTH
-                    ) * 10),
+                        MAX_PATH_LENGTH_PER_DOMAIN.get(
```
This seems unnecessary.

hartikainen commented:
This is starting to look good. One thing in addition to the above comments: could you also implement `SQL.tf_saveables`? Should look similar to what we have in `SAC.tf_saveables`. That should basically return a dict of all the tensorflow objects *created inside the algorithm class*.  For SQL, I think it would include at least the target Q, and all the optimizers.

bohan-zhang commented:
@hartikainen I implemented the `tf_saveables` for `SQL` and it seems to be working while training HalfCheetah! Let me know if there's anything else you want me to change 👍

haarnoja commented:
@bohan-zhang Thanks a lot for contributing this!! 

It could also make sense to include a second Q-function, similar to what we have in SAC. Basically the idea is to replace the Q-function with a pointwise minimum of two Q-functions, whenever it is used as a target network for something, and then learn these two Q-functions independently. @azhou42 I think you already implemented it in the private repo, could you merge the changes here too?

hartikainen commented:
@bohan-zhang I implemented the double Q function, refactored the SQL a bit, and did some testing with this. Seems like everything now *runs* fine, both in `examples.multi_goal.main` and `examples.development.main`. However, I'm not yet sure if the algorithm fully works. Looking at the multi goal example, the Q function looks good but the samples from the policy seem odd.

Still need to figure out what's wrong before we can merge this.

hartikainen commented on softlearning/algorithms/sql.py:
```diff
@@ -176,50 +190,82 @@ def _create_td_update(self):
 
         # \hat Q in Equation 11:
-        ys = tf.stop_gradient(
+        Q_target = tf.stop_gradient(
```
This could be moved to `self._get_Q_target` as is done in SAC.

bohan-zhang commented:
> @bohan-zhang I implemented the double Q function, refactored the SQL a bit, and did some testing with this. Seems like everything now _runs_ fine, both in `examples.multi_goal.main` and `examples.development.main`. However, I'm not yet sure if the algorithm fully works. Looking at the multi goal example, the Q function looks good but the samples from the policy seem odd.
> 
> Still need to figure out what's wrong before we can merge this.

Thank you for the update @hartikainen!

bohan-zhang commented:
@hartikainen I'm thinking about implementing Q-learning as a baseline to benchmark SAC and SQL. Can I add it as a new algorithm in `softlearning/algorithms`?

Also, do you have a rough estimate when we are going to merge this PR? Thank you!

hartikainen commented:
@bohan-zhang, as mentioned in #27, I was previously a bit skeptical about adding new baseline algorithms to this repo, because of the possible unnecessary overhead from maintaining them. However, I think we would be willing to give it a try and see if the overhead is small enough compared to the benefits of having more algorithms. So if you do want to implement standard Q-learning here, I'd say go for it! `softlearning/algorithms` would be the right place for that.

Re time estimate for this PR. I think this is ready to go right after we have verified that the SQL performance is roughly on par with the results in the [SAC paper](https://arxiv.org/pdf/1801.01290.pdf) at least in one of the mujoco environments, and also that the multi goal example works. We're working on ICML submission right now so I can't guarantee to have cycles before that deadline. If you have cycles, I think debugging the multigoal should at least be quite straightforward, and once that works well, I believe the other envs would work too.

bohan-zhang commented:
@hartikainen Thank you for the pointers!

Regarding adding Q-learning as a baseline, I'll give it a shot and see if it takes too much work.

Regarding debugging `SQL`, I am able to run everything as you mentioned, and in the `multi-goal` environment, after 100 iterations, the algorithm is able to explore multiple goals and get fairly close to them, so I'm not sure what I should try to debug here. I don't think I have enough knowledge to understand what you mean by the samples from the policy seem odd. I will run the newly fixed `SQL` on swimmer env and hopper env tonight and I'll share with you my results!

Good luck on your ICML submission!

hartikainen commented:
@bohan-zhang Sorry, I should've clarified what I mean by the policy seeming odd. In the plots that we render, the Q-function (the contours) looks what it should look like: for state `(0, 0)` we get 4 modes at the "cardinal directions" (up, right, down, left) towards the goals, for state `(2.5, 2.5)` we get 2 modes at the diagonals towards the goals `(5.0, 0.0)` and `(0.0, 5.0)`, etc. The blue samples in the figures, which present the actions that we have samples from the policy, on the other hand, did not match what I was expecting. Those samples should have shape similar to the Q contour (i.e. more samples towards the direction of high Q-values), yet what I observed was the actions samples were squashed in a very small box or something.

Now that I think of it, the problem might actually be in the action squashing. If I remember correctly, in the old SQL implementation, action squashing is still implemented as part of the algorithm, whereas nowadays, it's [handled by the policy class](https://github.com/rail-berkeley/softlearning/blob/master/softlearning/policies/gaussian_policy.py#L95). If so, we should remove the squashing from the algorithm side. I'm not 100% sure if this is the problem though, and it's also possible that the squashing doesn't even affect the multi goal environment since the action space might be between `(-1, 1)` already.

If you can post figures from the multi goal plots and swimmer/hopper results, that would be super useful!

bohan-zhang commented:
@hartikainen it seems that SQL has some problems when running in swimmer / hopper environment. I plotted the x position vs. time in this [ipython notebook](https://drive.google.com/open?id=1v3KJf_Zs9AynI1QZRyCsgeyGYQJCJZ8N)

There might be 2 reasons:
1. I didn't have enough iterations for SQL
2. SQL has bugs

Are there any other plots you are looking for?

bohan-zhang commented:
Thank you so much for investigating this, @hartikainen!

hartikainen commented:
Hey @bohan-zhang, I just spend a bit debugging this issue, and it seems like it's been solved now. There we couple of shapes that were flipped (see https://github.com/rail-berkeley/softlearning/pull/23/commits/05f86d99d84ea72b63d578f64be5f220156a975c).

The multi goal environment works well now. I'm gonna run couple of experiments with mujoco environments and merge this if things look good there.

hartikainen commented:
Okay, I ran some experiments with the recent fixes. It seems like the code works, even though the results are not quite on par with what we report in the original SAC paper. However, I'm inclined to believe this is due to hyperparameters and not a bug in the code.

Here are the results for reference:
### Hopper

### Walker

### HalfCheetah

### Ant

### Humanoid

hartikainen commented:
I'll merge this soon. Thanks a lot for the contribution @bohan-zhang! Very much appreciated!

bohan-zhang commented:
@hartikainen I am really glad that SQL will be merged soon, and it's been a pleasure working on this with you!
